starvingdead.lua

```diff
@@ -0,0 +1,128 @@
+-- Weaken and eventually destroy undead over time.
+--@enable = true
+--@module = true
+
+local argparse = require('argparse')
+local json = require('json')
+local persist = require('persist-table')
+
+local GLOBAL_KEY = 'starvingdead'
+
+StarvingDead = defclass(StarvingDead)
+StarvingDead.ATTRS{
+  decay_rate = 1,
+  death_threshold = 6
+}
+
+function StarvingDead:init()
+  self.timeout_id = nil
+  -- Percentage goal each attribute should reach before death.
+  self.attribute_goal = 10
+  self.attribute_decay = (self.attribute_goal ^ (1 / ((self.death_threshold * 28 / self.decay_rate)))) / 100
+  self.undead_count = 0
+
+  self:checkDecay()
+  print(([[StarvingDead started, checking every %s days and killing off at %s months]]):format(self.decay_rate, self.death_threshold))
+end
+
+function StarvingDead:checkDecay()
+  self.undead_count = 0
+  for _, unit in pairs(df.global.world.units.active) do
+    if (unit.enemy.undead and not unit.flags1.inactive) then
+      self.undead_count = self.undead_count + 1
+
+      -- time_on_site is measured in ticks, a month is 33600 ticks.
+      -- @see https://dwarffortresswiki.org/index.php/Time
+      for _, attribute in pairs(unit.body.physical_attrs) do
+        attribute.value = math.floor(attribute.value - (attribute.value * self.attribute_decay))
+      end
+
+      if unit.curse.time_on_site > (self.death_threshold * 33600) then
+        unit.flags1.inactive = true
+        unit.curse.rem_tags2.FIT_FOR_ANIMATION = true
+      end
+    end
+  end
+
+  self.timeout_id = dfhack.timeout(self.decay_rate, 'days', self:callback('checkDecay'))
+end
+
+if not dfhack.isMapLoaded() then
+  qerror('This script requires a fortress map to be loaded')
+end
```

Comment:
this needs to be after the isEnabled() global and onStateChange hook, otherwise those will fail to be found when DFHack starts and state will not get reloaded on map load